### ISSUE TITLE
Split CatalogPage's monolith code

### ISSUE THREAD

**Issue by duker33** [fidals/shopelectro#550]
shopelectro.views.catalog.CategoryPage currently consists of this steps:
- get products
- filter tags
- get template context
- paginate products

All these steps are semantically separated from CategoryPage and from each other.
Here is good code draft example for `get products` + `filter tags`.

```
class CategoryPage:

    def __init__(self, category_id):
        pass

    def products(self) -> 'QuerySet':
        return None

    def get_context_data(self):
        return {}


class TaggedCategoryPage:
    def __init__(self, page: CategoryPage, tags):
        self.super = page

    def products(self) -> 'QuerySet':
        return super.products.filter(smth=anyth)

    def get_context_data(self):
        context = self.super.get_context_data()
        data = {}
        return {
            'all_products': products,
            **data,
            **context,
        }
```

But use in this task traditional inheritance instead of objects composition

**Comment by artemiy312** [comment]
@duker33 Currently this issue does not look helpful and even vice versa, because we are going to create a complex inheritance structure, that will decrease view performance and will be still monolith. And in final we will use [only one class from this structure](https://github.com/fidals/shopelectro/pull/551/files#diff-99d29dab4b54555e71b666bfabb46ac1) that can be presented as single class as it is right now.

**Comment by duker33** [comment]
@artemiy312 , inheritance is only step for objects composition.

But even with inheritance we'll switch to named chunks of code with separated responsibility. Instead of one huge `get_context_method` with all-in-one.

With inheritance responsibility will be separated, but badly organized.
And it's task of objects composition to organize it well.
We'll enable composition after this task tree will fully 

```
class AbstractContext:
	def __init__(self, super: CatalogContext):
		self.super = super


class CatalogContext(AbstractContext):
	
    def products(catalog):
		return DB.fetch_prods(catalog)
		
	def get_context_data():
		return {...}  # something simple only about catalog

class TaggedCatalogContext:
    def products(tags):
		return self.super.products().and_aother_fetch(catalog)
		
	def get_context_data():
		return {
			**self.super.get_context_data(),
			... # and something simple only about tags
		}
		
class PaginatedCatalogContext:
	...
	
class DBTemplateContext:
	...

class CatalogView:
	def get_context_data(category_id, page_num):
		category = some_fetch(category_id)
		context = CatalogContext(category) | PaginatedCatalogContext(page_num)
		return context.get_context_data()

class TaggedCatalogView:
	def get_context_data(category_id, page_num, tags):
		category = some_fetch(category_id)
		context = (
			CatalogContext(category)
			| PaginatedCatalogContext(page_num)
			| TaggedCatalogContext(tags)
			| DBTemplateContext(tags)
		)
		return context.get_context_data()

```

In this case we can decrease amount of composited code for individual url

**Comment by duker33** [comment]
discussed this task with Artemiy via call.
I'll implement objects composition in this task. I'll increase hours count for it

**Comment by duker33** [comment]
broke pdd rules here. So increase time. Wasted it for arch

**Comment by 0pdd** [comment]
@duker33 4 puzzles [#566](https://github.com/fidals/shopelectro/issues/566), [#567](https://github.com/fidals/shopelectro/issues/567), [#568](https://github.com/fidals/shopelectro/issues/568), [#569](https://github.com/fidals/shopelectro/issues/569) are still not solved.

**Comment by 0pdd** [comment]
@duker33 7 puzzles [#566](https://github.com/fidals/shopelectro/issues/566), [#567](https://github.com/fidals/shopelectro/issues/567), [#568](https://github.com/fidals/shopelectro/issues/568), [#569](https://github.com/fidals/shopelectro/issues/569), [#577](https://github.com/fidals/shopelectro/issues/577), [#578](https://github.com/fidals/shopelectro/issues/578), [#579](https://github.com/fidals/shopelectro/issues/579) are still not solved.

**Comment by 0pdd** [comment]
@duker33 all 7 puzzles are solved here: [#566](https://github.com/fidals/shopelectro/issues/566), [#567](https://github.com/fidals/shopelectro/issues/567), [#568](https://github.com/fidals/shopelectro/issues/568), [#569](https://github.com/fidals/shopelectro/issues/569), [#577](https://github.com/fidals/shopelectro/issues/577), [#578](https://github.com/fidals/shopelectro/issues/578), [#579](https://github.com/fidals/shopelectro/issues/579).
